File: DiceRollerMultipleDices.py
__module_name__ = "Dice Roller Multiple"
__module_version__ = "0.1"
__module_description__ = "Rolador de Dados Multiplos em Python"

import hexchat
import re
from random import seed, randint

def rolador(comando, nome):
    regex = r"([+-]?)(\d+)(?:d(\d+))?"
    texto = "Rolei " + comando + " para " + str(nome) + " e os resultados foram ["

    padrao = re.findall(regex, comando)
    resultadoFinal = 0

    for sinal, quantidade, lados in padrao:
        quantidade = int(quantidade)
        sinalChar = sinal
        sinal = -1 if sinal == '-' else 1

        if lados:
            lados = int(lados)
            textoParte, valorParcial = rolar_dado(quantidade, lados)
            resultadoFinal += sinal * valorParcial
        else:
            valorParcial = quantidade
            textoParte = str(sinal * quantidade)
            resultadoFinal += sinal * quantidade

        texto += sinalChar+" (" + textoParte + ") "

    texto += "] = " + str(resultadoFinal) + ""
    return texto

def rolar_dado(qtd, lados):
    seed()
    resultadoFinal = 0
    texto = ""

    for i in range(1, qtd + 1):
        if i > 1:
            texto += " ; "

        resultado = randint(1, lados)
        resultadoFinal += resultado

        if resultado < (lados / 3):
            texto += "0,4 " + str(resultado) + " "
        elif resultado < ((lados / 3) * 2):
            texto += "0,7 " + str(resultado) + " "
        else:
            texto += "0,3 " + str(resultado) + " "

    return texto, resultadoFinal

def comandoRolador(word, word_eol, userdata):
    isOk = 0

    if len(word) == 2:
        nomePersonagem = hexchat.get_pluginpref("name_character")

        if nomePersonagem:
            nome = nomePersonagem
            comando = word[1]
            isOk = 1
        else:
            print("Você ainda não atribuiu um personagem. Por favor utilize /rolam OPERAÇÃO NOME_PERSONAGEM antes de prosseguir. Exemplo: /rolam 1d4+1d6 Juju")
    elif len(word) == 3:
        comando = word[1]
        nome = word[2]
        hexchat.set_pluginpref("name_character", nome)
        isOk = 1

    if isOk == 1:
        frase = rolador(comando, nome)
        canal = hexchat.get_info("channel")
        hexchat.command("msg " + canal + " " + frase)
    else: 
        print("Erro ao executar. Verifique!")

    return hexchat.EAT_ALL

hexchat.hook_command("rolam", comandoRolador, "Uso: /rolam OPERAÇÃO NOME_PERSONAGEM na primeira rolagem, após isso /rolam OPERAÇÃO.")

File: DiceRollerMultipleDices.lua
local nome = "Rolador4.0 - Lua Version"
local versao = 4
local desc = "Rolador de dados Multiplos"
math.randomseed(os.time())

local personagem_arquivo = hexchat.get_info("configdir") .. "/nome_personagem.txt"

local function salvar_personagem(nome)
    local f = io.open(personagem_arquivo, "w")
    if f then
        f:write(nome)
        f:close()
    end
end

local function carregar_personagem()
    local f = io.open(personagem_arquivo, "r")
    if f then
        local nome = f:read("*l")
        f:close()
        return nome
    end
    return nil
end


local function rolar_dado(qtd, lados)
    local total = 0
    local texto = ""

    for i = 1, qtd do
        if i > 1 then
            texto = texto .. " ; "
        end

        local resultado = math.random(1, lados)
        total = total + resultado

        if resultado < (lados / 3) then
            texto = texto .. "0,4 " .. resultado .. " "
        elseif resultado < ((lados / 3) * 2) then
            texto = texto .. "0,7 " .. resultado .. " "
        else
            texto = texto .. "0,3 " .. resultado .. " "
        end
    end

    return texto, total
end

local function rolador(expr, nome)
    local resultado_final = 0
    local texto = "Rolei " .. expr .. " para " .. nome .. " e os resultados foram ["

    for sinal, qtd, lados in string.gmatch(expr, "([+-]?)(%d+)[dD]?(%d*)") do
        local mult = (sinal == "-") and -1 or 1
        local quantidade = tonumber(qtd)
        local texto_parte = ""
        local valor = 0

        if lados ~= "" then
            lados = tonumber(lados)
            texto_parte, valor = rolar_dado(quantidade, lados)
        else
            valor = quantidade
            texto_parte = tostring(mult * quantidade)
        end

        resultado_final = resultado_final + (mult * valor)
        texto = texto .. sinal .. " (" .. texto_parte .. ") "
    end

    texto = texto .. "] = " .. tostring(resultado_final) .. ""
    return texto
end

local function comando_rolar(word, word_eol, userdata)
    comando = word[2]
    nome = word[3]

    if not comando then
        hexchat.print("Uso: /rolam EXPRESSAO [NOME_PERSONAGEM]")
        return hexchat.EAT_ALL
    end

    if not nome then
        nome = carregar_personagem()
        if not nome then
            hexchat.print("Você ainda não atribuiu um personagem. Ex: /rolam 1d6+1d4 Juju")
            return hexchat.EAT_ALL
        end
    else
        salvar_personagem(nome)
    end

    hexchat.print("O nome é "..nome)

    local frase = rolador(comando, nome)
    local canal = hexchat.get_info("channel")
    hexchat.command("msg " .. canal .. " " .. frase)

    return hexchat.EAT_ALL
end

hexchat.register(nome,versao,desc)
hexchat.hook_command("ROLAM",comando_rolar,desc)